File: services/geminiService.ts

import { GoogleGenAI, Type, Schema } from "@google/genai";
import { SunoResponse, GenerationMode, GroundingSource, SearchEngine } from "../types";
import { fileToGenerativePart } from "../utils";

const SYSTEM_INSTRUCTION = `
あなたはSuno.aiのエキスパートであり、最新のモデル（v4.5以降）を熟知した音楽プロデューサーです。
ユーザー入力を解析し、Suno.ai Custom Modeで最高の楽曲を生成するためのパラメータをJSON形式で出力してください。

## YouTube URLの処理（厳守事項）
1. 入力にYouTube URLがある場合、**Google Searchを使用して、その動画ID(v=...)に対応する正確な「動画タイトル」と「アーティスト/投稿者名」を特定してください。**
2. プレイリスト内の他の曲や、検索結果に出てくる「人気の関連動画（例：Vaundy, なとり等）」と混同しないでください。**動画IDに紐づく情報を100%優先してください。**
3. **「analysis」フィールドの冒頭に必ず以下の形式で記述してください：**
   「特定した動画：[動画タイトル] / 投稿者：[アーティスト名]」
4. もしタイトルが特定できない場合は、推測せず「動画の詳細を特定できませんでした」と記述してください。
5. **動画情報の後に、どのような音楽なのか（ジャンル、雰囲気、特徴など）を簡潔な文章で追記してください。**
   例：「... / 投稿者：Airis。疾走感のあるピアノとエモーショナルなボーカルが特徴的な、切ないJ-Popロックナンバーです。」

## 音楽解析のポイント
- 特定した楽曲の「ジャンル（例：Dark EDM, Phonk, J-Pop）」「BPM/テンポ感」「使用されている特徴的な楽器」「ボーカルの雰囲気（中性的、ダウナー、力強い等）」を分析してください。
- Suno v4.5の特性（高音質、複雑な構成の理解）を活かしたプロンプトを作成してください。

## 構成ルール
- [Intro], [Verse], [Chorus], [Bridge], [Drop], [Outro] などのメタタグを効果的に使用。
- 歌詞は動画の世界観を踏襲し、かつSunoで生成した際に中毒性が出るように構成。

## 禁止事項とアレンジルール（最重要）
1. **「タイトル」および「歌詞」の完全コピーは厳禁です。** 元の楽曲や動画の要素をそのまま出力しないでください。
2. YouTube動画を解析する場合、動画の概要欄、コメント欄、背景情報、投稿者のマインドなどを深く読み取り、**必ず独自の解釈やアレンジを加えてください。**
3. 元の世界観やメッセージ性は尊重しつつ、Suno.aiで生成した際に「新しい魅力」が生まれるように、言葉選びや構成にオリジナリティを持たせてください。

## タイトル候補のルール（厳守）
- タイトル候補（titleCandidates）は必ず「日本語 / English」の形式で出力してください。
- 例: 「夜に駆ける / Racing Into The Night」「電脳都市 / Cybernetic City」
`;

const selectionSchema = {
  type: Type.OBJECT,
  properties: {
    title: { type: Type.STRING, description: "Bilingual format: 日本語タイトル / English Title" },
    style: { type: Type.STRING },
    instrumental: { type: Type.BOOLEAN },
    content: { type: Type.STRING, description: "Full lyrics or structure using v4.5 metatags." },
    comment: { type: Type.STRING, description: "Brief explanation of the musical choices." },
  },
  required: ["title", "style", "instrumental", "content", "comment"],
};

const responseSchema: Schema = {
  type: Type.OBJECT,
  properties: {
    analysis: { type: Type.STRING, description: "Must start with identified Video Title/Creator, followed by a brief musical description." },
    titleCandidates: {
      type: Type.ARRAY,
      items: { type: Type.STRING },
      description: "Exactly 5 bilingual titles in format: 日本語 / English",
    },
    styleCandidates: {
      type: Type.ARRAY,
      items: { type: Type.STRING },
      description: "Exactly 5 different style prompts (English tags only).",
    },
    bestSelection: { ...selectionSchema },
    alternativeSelection: { ...selectionSchema },
  },
  required: ["analysis", "titleCandidates", "styleCandidates", "bestSelection", "alternativeSelection"],
};

/**
 * Supported oEmbed providers and their patterns
 */
const OEMBED_PROVIDERS = [
  {
    name: 'YouTube',
    patterns: [/youtube\.com\/watch\?v=/, /youtu\.be\//, /youtube\.com\/embed\//, /youtube\.com\/v\//],
    endpoint: (url: string) => `https://www.youtube.com/oembed?url=${encodeURIComponent(url)}&format=json`
  },
  {
    name: 'Spotify',
    patterns: [/open\.spotify\.com\/(track|album|playlist|artist)\//],
    endpoint: (url: string) => `https://open.spotify.com/oembed?url=${encodeURIComponent(url)}`
  },
  {
    name: 'SoundCloud',
    patterns: [/soundcloud\.com\//],
    endpoint: (url: string) => `https://soundcloud.com/oembed?url=${encodeURIComponent(url)}&format=json`
  },
  {
    name: 'TikTok',
    patterns: [/tiktok\.com\/.*\/video\//, /tiktok\.com\/t\//],
    endpoint: (url: string) => `https://www.tiktok.com/oembed?url=${encodeURIComponent(url)}`
  },
  {
    name: 'Vimeo',
    patterns: [/vimeo\.com\//],
    endpoint: (url: string) => `https://vimeo.com/api/oembed.json?url=${encodeURIComponent(url)}`
  },
  {
    name: 'Twitter/X',
    patterns: [/twitter\.com\//, /x\.com\//],
    endpoint: (url: string) => `https://publish.twitter.com/oembed?url=${encodeURIComponent(url)}`
  },
  {
    name: 'Instagram',
    patterns: [/instagram\.com\//],
    endpoint: (url: string) => `https://api.instagram.com/oembed?url=${encodeURIComponent(url)}`
  }
];

/**
 * Fetch metadata from various URLs using oEmbed
 */
const fetchUrlMetadata = async (url: string): Promise<{ title: string; author: string; provider: string } | null> => {
  if (!url) return null;

  const provider = OEMBED_PROVIDERS.find(p => p.patterns.some(pattern => pattern.test(url)));
  if (!provider) return null;

  try {
    const response = await fetch(provider.endpoint(url));
    if (!response.ok) return null;
    const data = await response.json();
    return {
      title: data.title || '',
      author: data.author_name || data.author || '',
      provider: provider.name
    };
  } catch (error) {
    console.error(`oEmbed fetch error for ${provider.name}:`, error);
    return null;
  }
};

/**
 * Search using Google Custom Search API
 */
const searchWithGoogleCustom = async (query: string): Promise<string> => {
  const apiKey = (process.env as any).GOOGLE_CUSTOM_SEARCH_API_KEY;
  const cx = (process.env as any).GOOGLE_CUSTOM_SEARCH_CX;

  if (!apiKey || !cx) {
    return "Google Custom Search API key or CX not configured. Please set GOOGLE_CUSTOM_SEARCH_API_KEY and GOOGLE_CUSTOM_SEARCH_CX in .env.local.";
  }

  try {
    const url = `https://www.googleapis.com/customsearch/v1?key=${apiKey}&cx=${cx}&q=${encodeURIComponent(query)}&num=5`;
    const response = await fetch(url);
    const data = await response.json();

    if (data.items && data.items.length > 0) {
      return data.items.map((item: any) => `- ${item.title}: ${item.snippet}`).join('\n');
    }
    return "No results found.";
  } catch (error) {
    console.error("Google Custom Search error:", error);
    return "Search failed.";
  }
};

/**
 * Search using Tavily API
 */
const searchWithTavily = async (query: string): Promise<string> => {
  const apiKey = (process.env as any).TAVILY_API_KEY;

  if (!apiKey) {
    return "Tavily API key not configured. Please set TAVILY_API_KEY in .env.local.";
  }

  try {
    const response = await fetch('https://api.tavily.com/search', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        api_key: apiKey,
        query: query,
        search_depth: 'basic',
        max_results: 5
      })
    });
    const data = await response.json();

    if (data.results && data.results.length > 0) {
      return data.results.map((r: any) => `- ${r.title}: ${r.content}`).join('\n');
    }
    return "No results found.";
  } catch (error) {
    console.error("Tavily search error:", error);
    return "Search failed.";
  }
};

export const generateSunoPrompt = async (
  text: string,
  youtubeUrl: string,
  file: File | null,
  mode: GenerationMode = GenerationMode.AUTO,
  options: { searchEngine: SearchEngine; modelName: string; enableVideoAnalysis?: boolean } = { searchEngine: 'google-grounding', modelName: 'gemini-3-flash-preview', enableVideoAnalysis: false }
): Promise<SunoResponse> => {
  try {
    const apiKey = process.env.API_KEY;
    if (!apiKey) {
      throw new Error("API Key is missing.");
    }

    const genAI = new GoogleGenAI({ apiKey: apiKey });
    const modelName = options.modelName;

    const parts: any[] = [];
    if (file) {
      const mediaPart = await fileToGenerativePart(file);
      parts.push(mediaPart);
    }

    let prompt = `Current Generation Mode: ${mode}. \n`;
    let externalSearchResults = '';

    // Check if we have a URL to analyze
    const urlMetadata = youtubeUrl ? await fetchUrlMetadata(youtubeUrl) : null;

    if (urlMetadata) {
      const isVideoAnalysisEnabled = options.enableVideoAnalysis;

      prompt += `=== CRITICAL: EXTERNAL MEDIA IDENTIFICATION ===\n`;
      prompt += `Source URL: ${youtubeUrl}\n`;
      prompt += `Platform: ${urlMetadata.provider}\n`;
      prompt += `✅ VERIFIED MEDIA INFORMATION:\n`;
      prompt += `  - Title: ${urlMetadata.title}\n`;
      prompt += `  - Author/Artist: ${urlMetadata.author}\n\n`;

      prompt += `INSTRUCTION: Analyze this ${urlMetadata.provider} content. Use the verified title and author above.\n`;
      prompt += `Do NOT guess the content. If you cannot analyze it, stay generic but professional.\n\n`;

      if (isVideoAnalysisEnabled && urlMetadata.provider === 'YouTube') {
        prompt += `VIDEO ANALYSIS MODE: Research and analyze the visual and auditory style of this video.\n\n`;
      }

      // Additional search for context (genre, style, etc.)
      const searchQuery = `"${urlMetadata.title}" ${urlMetadata.author} music genre style analysis`;

      switch (options.searchEngine) {
        case 'google-grounding':
          prompt += `OPTIONAL: Use Google Search to find more about the musical style of "${urlMetadata.title}".\n`;
          break;
        case 'google-custom':
          externalSearchResults = await searchWithGoogleCustom(searchQuery);
          prompt += `=== ADDITIONAL CONTEXT ===\n${externalSearchResults}\n`;
          break;
        case 'tavily':
          externalSearchResults = await searchWithTavily(searchQuery);
          prompt += `=== ADDITIONAL CONTEXT ===\n${externalSearchResults}\n`;
          break;
      }
    } else if (youtubeUrl) {
      // Fallback for non-oEmbed URLs (like Suno.ai)
      const isSuno = youtubeUrl.includes('suno.com') || youtubeUrl.includes('suno.ai');

      prompt += `=== MEDIA ANALYSIS (NON-OEMBED) ===\n`;
      prompt += `Target URL: ${youtubeUrl}\n`;

      if (isSuno) {
        prompt += `Platform: Suno.ai\n`;
        prompt += `INSTRUCTION: This is a Suno.ai song link. Please use your search capability (Google Search) to find the title, tags, and lyrics of this track if available.\n`;
      } else {
        prompt += `Note: Could not fetch automatic metadata via oEmbed.\n`;
      }

      prompt += `INSTRUCTION: Analyze the contents of this URL to provide a high-quality music prompt.\n\n`;

      // Trigger search if enabled
      if (options.searchEngine === 'google-grounding') {
        prompt += `SEARCH INSTRUCTION: Search for this specific URL: ${youtubeUrl} to extract track details.\n`;
      }
    }

    if (text) {
      prompt += `Theme/Concept: ${text}\n`;
    } else if (!file) {
      prompt += "Create a high-quality, modern music concept.";
    }

    parts.push({ text: prompt });

    const config: any = {
      systemInstruction: SYSTEM_INSTRUCTION,
      responseMimeType: "application/json",
      responseSchema: responseSchema,
      temperature: 0.1,
      topP: 0.1,
    };

    // Only use built-in Google Search for google-grounding
    if (options.searchEngine === 'google-grounding') {
      config.tools = [{ googleSearch: {} }];
    }

    const result = await genAI.models.generateContent({
      model: modelName,
      contents: { parts: parts },
      config: config,
    });

    if (!result.text) {
      throw new Error("AI returned an empty response.");
    }

    const jsonResponse = JSON.parse(result.text) as SunoResponse;

    // Extract grounding sources (only available for google-grounding)
    const sources: GroundingSource[] = [];
    const chunks = result.candidates?.[0]?.groundingMetadata?.groundingChunks;
    if (chunks) {
      chunks.forEach((chunk: any) => {
        if (chunk.web && chunk.web.uri) {
          sources.push({
            uri: chunk.web.uri,
            title: chunk.web.title || "Reference Source"
          });
        }
      });
    }

    const uniqueSources = Array.from(new Map(sources.map(s => [s.uri, s])).values());

    return {
      ...jsonResponse,
      sources: uniqueSources.length > 0 ? uniqueSources : undefined
    };
  } catch (error: any) {
    console.error("Gemini API Error:", error);
    throw error;
  }
};

File: server/gemini_service.py
import os
import json
import logging
import requests
import google.generativeai as genai
from typing import Optional, List, Dict, Any
from core.config import settings

logger = logging.getLogger("SunoArchitect.GeminiService")

# Configure Gemini
if settings.gemini_api_key:
    genai.configure(api_key=settings.gemini_api_key)

SYSTEM_INSTRUCTION = """
あなたはSuno.aiのエキスパートであり、最新のモデル（v4.5以降）を熟知した音楽プロデューサーです。
現在、ユーザーはAPIクォータ制限に直面しています。効率的かつ高品質なプロンプトを作成してください。

## 重要: 出力言語に関する絶対的なルール
- **すべてのテキスト出力（analysis, comment, titleの日本語部分など）は、必ず「日本語」で記述してください。**
- **英語での解説や分析は禁止です。**
- ただし、スタイルプロンプト（style）や歌詞の英単語などは英語のままで構いません。

## 解析の優先順位
1. **提供されたURLのメタデータ（タイトル、アーティスト、説明文）を最優先で分析対象とします。**これらはAPIを使わずに取得できる情報です。
2. 検索エンジン（Google Search等）の使用は、ユーザーが明示的に許可した場合に限ります。URLがある場合は、そのURLから推測される情報を「正解」として扱ってください。

## 音楽解析とプロンプト生成のポイント（Suno & ACE-Step対応）
- 特定した楽曲の「ジャンル（例：Dark EDM, Phonk, J-Pop）」「BPM/テンポ感」「使用されている特徴的な楽器」「ボーカルの雰囲気」「全体的なムード」をメタデータから推測・分析してください。
- Suno v4.5（カンマ区切りのタグに強い）と、**ACE-Step v1.5（自然言語での詳細な情景描写に強い）**の両方の特性を活かしたプロンプトを作成してください。
- **ACE-Step v1.5向けのスタイルプロンプトは、単なるカンマ区切りではなく、完全な英語の文章（2〜3文）による詳細な楽曲描写（テンポ、楽器構成、ボーカル音質、ムードの文脈的説明）が非常に有効です。**（例："An energetic J-pop track driven by rapid piano arpeggios that create a constant sense of motion alongside a punchy electronic drum beat..."）

## 構成ルール
- [Intro], [Verse], [Chorus], [Bridge], [Drop], [Outro] などのメタタグを使用。
- 歌詞は元の世界観を尊重しつつ、独自のアレンジを加えてください（完全コピー禁止）。

## タイトル候補
- **「日本語タイトル / English Title」** の形式で出力してください。必ず日本語と英語を併記してください。
"""

OEMBED_PROVIDERS = [
    {"name": "YouTube", "patterns": ["youtube.com", "youtu.be"], "endpoint": lambda url: f"https://www.youtube.com/oembed?url={url}&format=json"},
    {"name": "Spotify", "patterns": ["spotify.com"], "endpoint": lambda url: f"https://open.spotify.com/oembed?url={url}"},
    {"name": "SoundCloud", "patterns": ["soundcloud.com"], "endpoint": lambda url: f"https://soundcloud.com/oembed?url={url}&format=json"},
]

def fetch_url_metadata(url: str) -> Optional[Dict[str, Any]]:
    if not url:
        return None

    if "suno.com" in url or "suno.ai" in url:
        try:
            # Internal call to suno analyze
            from routers.suno import analyze_suno_logic
            data = analyze_suno_logic(url)
            if data:
                return {
                    "title": data.get("title"),
                    "author": "Suno User",
                    "provider": "Suno",
                    "description": data.get("description")
                }
        except Exception as e:
            logger.error(f"Suno analysis failed: {e}")

    for provider in OEMBED_PROVIDERS:
        if any(pattern in url for pattern in provider["patterns"]):
            try:
                response = requests.get(provider["endpoint"](url), timeout=5)
                if response.ok:
                    data = response.json()
                    return {
                        "title": data.get("title", ""),
                        "author": data.get("author_name", data.get("author", "")),
                        "provider": provider["name"]
                    }
            except Exception as e:
                logger.error(f"OEmbed fetch failed for {provider['name']}: {e}")
    
    return {"title": "Unknown URL", "author": "Unknown", "provider": "Web Link", "description": url}

async def generate_suno_prompt(
    text: str,
    youtube_url: str = "",
    image_path: Optional[str] = None,
    mode: str = "auto",
    options: Dict[str, Any] = None,
    theme: str = ""
) -> Dict[str, Any]:
    if options is None:
        options = {"searchEngine": "none", "modelName": "gemini-1.5-flash", "lyricsLanguage": "Japanese"}
    
    model_name = options.get("modelName", "gemini-1.5-flash")
    # Map old names if necessary
    if "gemini-3" in model_name:
        model_name = "gemini-1.5-pro" # Fallback to stable for now if 3 is not available in SDK yet

    model = genai.GenerativeModel(
        model_name=model_name,
        system_instruction=SYSTEM_INSTRUCTION
    )

    prompt_parts = []
    
    if image_path and os.path.exists(image_path):
        import PIL.Image
        img = PIL.Image.open(image_path)
        prompt_parts.append(img)
    
    meta_prompt = f"Current Generation Mode: {mode}. \n"
    url_metadata = fetch_url_metadata(youtube_url) if youtube_url else None

    if url_metadata:
        meta_prompt += f"=== MEDIA INFO ===\nTitle: {url_metadata['title']}\nArtist: {url_metadata['author']}\nSource: {url_metadata['provider']}\n"
        if url_metadata.get("description"):
            meta_prompt += f"Context: {url_metadata['description']}\n"
        meta_prompt += "\nINSTRUCTION: Analyze this media information to determine the musical style, genre, and lyrical themes. \n"

    if image_path:
        meta_prompt += "\nINSTRUCTION: Analyze the uploaded image. Focus on the **visual mood, atmosphere, and lighting** to determine the musical style.\n"

    if theme:
        meta_prompt += f"\nUser Provided Theme/Concept: \"{theme}\"\nIMPORTANT: Incorporate this theme into the lyrics and style analysis.\n"

    if text:
        meta_prompt += f"\nCurrent Prompt Context: {text}\n"

    if options.get("lyricsLanguage") == "English":
        meta_prompt += "\n[言語指定] 歌詞はすべて「英語（English）」で記述してください。日本語は使用しないでください。\n"
    else:
        meta_prompt += "\n[言語指定] 歌詞のベースは「日本語」としてください。\n"

    prompt_parts.append(meta_prompt)

    generation_config = {
        "temperature": 0.2,
        "response_mime_type": "application/json"
    }

    import asyncio
    max_retries = 3
    for attempt in range(max_retries):
        try:
            response = await model.generate_content_async(
                prompt_parts,
                generation_config=generation_config
            )
            return json.loads(response.text)
        except json.JSONDecodeError as e:
            logger.warning(f"JSON parse error on attempt {attempt + 1}: {e}. Retrying...")
            if attempt == max_retries - 1:
                logger.error(f"Failed to parse Gemini response after {max_retries} attempts")
                return {"error": "Failed to parse AI response", "raw": response.text if 'response' in locals() else ""}
            await asyncio.sleep(1)
        except Exception as e:
            logger.error(f"Gemini API error on attempt {attempt + 1}: {e}")
            if attempt == max_retries - 1:
                raise
            await asyncio.sleep(2)
    
    return {"error": "Failed after max retries"}

async def structure_lyrics(
    raw_lyrics: str,
    language: str = "ja",
    model_name: str = "gemini-1.5-flash"
) -> str:
    # Port structure_lyrics logic here
    model = genai.GenerativeModel(model_name=model_name)
    
    instruction = """You are a lyrics formatter specialized for Suno AI.
Your task is to structure raw lyrics into a format optimized for music generation.
Output ONLY the tagged lyrics. No explanations.
"""
    
    prompt = f"{instruction}\n\nLyrics to format:\n{raw_lyrics}"
    
    response = await model.generate_content_async(prompt)
    return response.text.strip()

async def generate_from_selected_title(
    selected_title: str,
    original_analysis: str,
    style_candidates: List[str],
    options: Dict[str, Any] = None
) -> Dict[str, Any]:
    if options is None:
        options = {"modelName": "gemini-1.5-flash", "lyricsLanguage": "Japanese"}
    
    model_name = options.get("modelName", "gemini-1.5-flash")
    model = genai.GenerativeModel(
        model_name=model_name,
        system_instruction=SYSTEM_INSTRUCTION
    )

    prompt = f"""
## タスク
以下の分析結果と選択されたタイトルに基づいて、Suno v4.5用の楽曲プロンプトを2パターン生成してください。

## 分析結果
{original_analysis}

## 選択されたタイトル
{selected_title}

## スタイル候補（参考）
{"\n".join(style_candidates)}

## 出力ルール
- bestSelection: 選択されたタイトルに最も適したスタイルと歌詞を生成
- alternativeSelection: 同じタイトルで異なるアプローチ（変化球）を提案
- commentは必ず日本語で記述すること
"""
    if options.get("lyricsLanguage") == "English":
        prompt += "\n[言語指定] 歌詞はすべて「英語（English）」で記述してください。日本語は使用しないでください。\n"
    else:
        prompt += "\n[言語指定] 歌詞のベースは「日本語」としてください。\n"

    generation_config = {
        "temperature": 0.3,
        "response_mime_type": "application/json"
    }

    import asyncio
    max_retries = 3
    for attempt in range(max_retries):
        try:
            response = await model.generate_content_async(
                prompt,
                generation_config=generation_config
            )
            data = json.loads(response.text)
            data["bestSelection"]["title"] = selected_title
            data["alternativeSelection"]["title"] = selected_title
            return data
        except json.JSONDecodeError as e:
            logger.warning(f"Phase 2 JSON parse error on attempt {attempt + 1}: {e}. Retrying...")
            if attempt == max_retries - 1:
                logger.error(f"Failed to parse Phase 2 response after {max_retries} attempts")
                return {"error": "Failed to parse AI response", "raw": response.text if 'response' in locals() else ""}
            await asyncio.sleep(1)
        except Exception as e:
            logger.error(f"Phase 2 API error on attempt {attempt + 1}: {e}")
            if attempt == max_retries - 1:
                raise
            await asyncio.sleep(2)
            
    return {"error": "Failed after max retries"}

async def generate_title(
    lyrics: str,
    theme: str,
    prompt: str,
    model_name: str = "gemini-1.5-flash"
) -> str:
    model = genai.GenerativeModel(model_name=model_name)
    
    instruction = """
        You are a visionary music producer and poet.
        Your task is to create a compelling, bilingual title for a song based on its lyrics, theme, and musical style.
        Return ONLY the title in "Japanese Title / English Title" format.
    """
    
    user_content = f"Theme: {theme}\nStyle/Prompt: {prompt}\nLyrics: {lyrics[:1000]}"
    
    response = await model.generate_content_async(f"{instruction}\n\n{user_content}")
    return response.text.strip()

async def generate_style_from_lyrics(
    lyrics: str,
    url: str = "",
    theme: str = "",
    language: str = "ja",
    model_name: str = "gemini-1.5-flash"
) -> str:
    model = genai.GenerativeModel(model_name=model_name)
    prompt = f"Generate a Suno style prompt (comma-separated tags) for these lyrics:\n{lyrics[:1000]}\nTheme: {theme}\nURL Info: {url}"
    response = await model.generate_content_async(prompt)
    return response.text.strip()

async def llm_proxy(body: Dict[str, Any]) -> Dict[str, Any]:
    """
    Proxies LLM requests to Gemini.
    """
    model_name = body.get("model", "gemini-1.5-flash")
    messages = body.get("messages", [])
    system_instruction = body.get("system_instruction", SYSTEM_INSTRUCTION)
    
    # Simple proxy implementation
    model = genai.GenerativeModel(
        model_name=model_name,
        system_instruction=system_instruction
    )
    
    # Convert message format if needed (Gemini expects history)
    history = []
    current_message = ""
    for msg in messages:
        if msg["role"] == "user":
            current_message = msg["content"]
        elif msg["role"] == "assistant":
            history.append({"role": "user", "parts": [current_message]})
            history.append({"role": "model", "parts": [msg["content"]]})
            current_message = ""
            
    if current_message:
        chat = model.start_chat(history=history)
        response = await chat.send_message_async(current_message)
        return {"content": response.text}
    
    return {"error": "No user message found"}
